core/iwasm/libraries/libc-builtin/libc_builtin_wrapper.c

```diff
@@ -341,15 +216,29 @@ _vprintf_wa(out_func_t out, void *ctx, const char *fmt, _va_list ap,
 
                     s = start = addr_app_to_native(s_offset);
 
-                    while (*s)
-                        out((int)(*s++), ctx);
+                    str_len = strlen(start);
+                    if (str_len >= UINT32_MAX - 64) {
+                        print_err(out, ctx);
+                        break;
+                    }
+
+                    /* reserve 64 more bytes as there may be width description
+                     * in the fmt */
+                    buf_len = str_len + 64;
 
-                    if (padding == PAD_SPACE_AFTER) {
-                        int remaining = min_width - (int32)(s - start);
-                        while (remaining-- > 0) {
-                            out(' ', ctx);
-                        }
+                    if (!(buf = wasm_runtime_malloc(buf_len))) {
+                        print_err(out, ctx);
+                        break;
                     }
+
+                    memset(temp_fmt, 0, sizeof(temp_fmt));
+                    bh_memcpy_s(temp_fmt, sizeof(temp_fmt), fmt_start_addr,
+                                fmt - fmt_start_addr + 1);
+                    snprintf(buf, buf_len, temp_fmt, start);
```

Comment:
The result of printing NULL string is different from native, for case below:
```C
printf("Null string: %7s\n", NULL);
```
The output here is: `Null string:`
The output of native is: `Null string:  (null)`
Wondering whether we should check the NULL string in wasm mode, for example, here if s_offset is 0 and str_len is 0, then we treat the string as NULL:
```C
snprintf(buf, buf_len, temp_fmt, (s_offset == 0 && str_len == 0) ? NULL : start);
```